[FEATURE] Allow nested filters

e0ipso opened the pull request:
Support JSON API relational filters specification in
https://gist.github.com/mateu-aguilo-bosch/1acbced683908ff631e3

Commit:
Use the resource path

Commit:
[BUGFIX] (JSON API) Fix nested includes

Nested includes are now only included if requested. A included
compound document will be requested with dot notation, and RESTful
will ensure that full document linkage is provided.

Commit:
[FEATURE] (JSON API) Test request fix query

Add testing to the parent inclusion of the fields and filters.

Commit:
Fix tests with a nested entityreference

Commit:
Add testing for nested resources

Commit:
[FEATURE] Allow files to be references

Allow files to be treated as entity references. Instead of getting
the file array, get the entity id or the full entity. To use this
behaviour use the 'class' property in the publicFields() definition.

Commit:
[BUGFIX] Fix includes and fields for nested

Pass the request down to the resource fields so we can pass relevant
query string parameters to compound documents. The parameters are
detected by stripping the public name + a dot from the beginnining
of the parameter. So if 'related.images.uri' is passed to the top
request, 'images.uri' will be passed to the ResourceFieldResource
for 'related'. The resource field for 'images' will recieve 'uri'
in the generated request object to compound the document.

Commit:
[BUGFIX] Make getImageUri static function

So it can be referenced as process callbacks.

Commit:
[FEATURE] Adds base decorators for DataProviders

When a resource wants to modify something of the data provider, they
can bind to a data provider decorator. That custom decorator can
have public methods overridden.

Commit:
Fix the request passing to the resource fields

Commit:
[BUGFIX] Do not trigger access tag on sort

When there is a sort on a field (and therefor a join to the field
table) we cannot add the access tag. Just like we are skipping it
when there is a field condition.

Commit:
[FEATURE] Filter by the referencedIdProperty value

When outputting reference values you can use referencedPropertyId to
return a different field than the referenced entity's ID. With this
PR filters accept the same value that was returned. That allows you
for instance to filter based on taxonomy term name.

Fixes #598

Commit:
Fix user tests & add set and get for resourcePath

Commit:
Fix list view entity test

Commit:
Fix authentication tests

Commit:
Temporarily run conflicting tests

Commit:
Fix URLs and revert travis

Commit:
[FEATURE] Add tests to alternative reference id

Add basic testing to the reference fields to produce uuids.

Commit:
Filter for comparison

Commit:
Add more testing

Commit:
Add reference UUID testing and random comment fix.

Commit:
Assert the links and fix formatter's resource

Commit:
[FEATURE] Add testing to sparse fieldsets

Test that only the selected fields are included in the response for
nested resources.

Commit:
[FEATURE] Allow nested filters

Support JSON API relational filters specification in
https://gist.github.com/mateu-aguilo-bosch/1acbced683908ff631e3

e0ipso commented:
Depends on #580.

Commit:
Merge branch '7.x-2.x' into 2.x-relational-filters

Conflicts:
	src/Plugin/resource/DataProvider/DataProvider.php

Commit:
UNDO this commit

Commit:
Commit undone

Commit:
Remove test concurrency

Commit:
Fix error detection in Travis

Commit:
Set interpreter for filter evaluation

e0ipso commented:
Green!

amitaibu commented on src/Plugin/resource/DataProvider/DataProviderDecorator.php:
```diff
@@ -0,0 +1,186 @@
+use Drupal\restful\Http\RequestInterface;
+use Drupal\restful\Plugin\resource\Field\ResourceFieldInterface;
+
+abstract class DataProviderDecorator implements DataProviderInterface {
```
same idea as a resource decorator? the use case here is to alter the query?

e0ipso replied:
Yep. See https://github.com/RESTful-Drupal/restful/pull/616/files#r38285763.

amitaibu commented on src/Util/EntityFieldQuery.php:
```diff
@@ -0,0 +1,137 @@
+      // for a condition at the end.
+      // Relationships start with the entity base table.
+      $entity_info = entity_get_info($this->entityConditions['entity_type']['value']);
+      $entity_table = $entity_table_alias = $entity_info['base table'];
```
the `base table` might not be the entity table - it might be the field table -- it depends on the order the query was built. I Had the unfortunate opportunity to [deal](https://github.com/Gizra/og/blob/7.x-2.x/og.module#L1370) with how an alterable EFQ for D7 is on OG ;)

amitaibu commented on src/Util/EntityFieldQuery.php:
```diff
@@ -0,0 +1,137 @@
+  /**
+   * {@inheritdoc}
+   */
+  public function finishQuery($select_query, $id_key = 'entity_id') {
```
this code is awesome and horrible at the same time :P. EFQ makes it so hard to alter it without prior knowledge of the actually query. You are a brave person :)

e0ipso replied:
It's major brain damage trying to figure this out.

amitaibu commented on src/Util/EntityFieldQuery.php:
```diff
@@ -0,0 +1,137 @@
+   * @return string
+   *   The table alias.
+   */
+  protected static function aliasJoinTable($table_name, \SelectQuery $query) {
```
I think you can check it with the join type like [here](https://github.com/Gizra/og/blob/7.x-2.x/og.module#L1389)

e0ipso replied:
Actually this is only making sure we generate unique names for our alias that can be traced back to the aliased table easily.

amitaibu commented on src/Util/RelationalFilter.php:
```diff
@@ -0,0 +1,115 @@
+  /**
+   * {@inheritdoc}
+   */
+  public function isIsDestination() {
```
wrong name?

e0ipso replied:
Removed.

:+1:

amitaibu commented on src/Util/RelationalFilterInterface.php:
```diff
@@ -0,0 +1,60 @@
+  public function getType();
+
+  /**
+   * Is this the last relational filter?.
```
why do we need this? Also, the name doesn't explain well what it is doing.

e0ipso replied:
We don't.

:+1:

amitaibu commented on src/Http/Request.php:
```diff
@@ -320,11 +321,42 @@ protected static function parseBodyContentType($content_type) {
+   *   The parsed input array.
+   */
+  protected function fixQueryFields(array $input) {
+    // Make sure to add all of the parents for the dot-notation sparse
```
move this to the PHPdocs.

amitaibu commented on src/Plugin/formatter/FormatterJsonApi.php:
```diff
@@ -153,7 +143,18 @@ protected function extractFieldValues($data, &$included) {
+          // including the parents.
+
+          // Remove numeric parents since those only indicate that the field was
+          // multivalue, not a parent.
```
worth an example

amitaibu commented on src/Plugin/formatter/FormatterJsonApi.php:
```diff
@@ -251,4 +252,32 @@ protected function getRequest() {
   }
 
+  /**
+   * Move the embedded resources to the included key.
```
so we always want an `included` key for embeds, also for the simple json formatter?

I'm actually wondering if we need/want to keep the simple json, and default to json api? (out of scope of this PR of course)

amitaibu replied:
actually, maybe for BC we should keep the simple json as is?

e0ipso replied:
This file is the JSON API formatter. Anything in this file only applies to that.

amitaibu commented on src/Plugin/resource/DataProvider/CacheDecoratedDataProvider.php:
```diff
@@ -46,6 +46,20 @@ public function __construct(DataProviderInterface $subject, \DrupalCacheInterfac
   /**
    * {@inheritdoc}
    */
+  public static function isNestedField($field_name) {
```
why not extend `DataProvider` and begone with this duplicated proxy methods?

e0ipso replied:
So you can inherit from different parents, and decide it at runtime.

If you need to alter (for instance) `entityCheckAccess` for your resource you need to:
- Override the `Resource::dataProviderClassName` to return a custom data provider class name `\Drupal\…\DataProviderNoAccessCheck`.
- Implement the new `class DataProviderNoAccessCheck extends DataProviderNode` and override `entityCheckAccess`.

At that point, you cannot add caching to that since your class extends from `DataProviderNode`, and not the hypothetical `class DataProviderCached extends DataProviderBase`.

amitaibu commented on src/Plugin/resource/DataProvider/DataProvider.php:
```diff
@@ -208,7 +254,8 @@ public function canonicalPath($path) {
    */
   public function methodAccess(ResourceFieldInterface $resource_field) {
-    return in_array($this->getRequest()->getMethod(), $resource_field->getMethods());
+    return in_array($this->getRequest()
```
was more readable the previous way.

e0ipso replied:
PhpStorm's auto-fix feature…

e0ipso replied:
:+1:

amitaibu commented on src/Plugin/resource/DataProvider/DataProviderEntity.php:
```diff
@@ -44,7 +50,7 @@ class DataProviderEntity extends DataProvider implements DataProviderEntityInter
    * @var string
    */
-  protected $EFQClass = '\EntityFieldQuery';
+  protected $EFQClass = '\Drupal\restful\Util\EntityFieldQuery';
```
nice! :)

e0ipso replied:
abstraction++

amitaibu commented on src/Plugin/resource/DataProvider/DataProviderEntity.php:
```diff
@@ -592,28 +598,73 @@ protected function queryForListSort(\EntityFieldQuery $query) {
         for ($index = 0; $index < count($filter['value']); $index++) {
-          $query->fieldCondition($property_name, $resource_field->getColumn(), $filter['value'][$index], $filter['operator'][$index]);
+          // If referencing an entity by an alternate ID, change the user
+          // provided value by the entity value.
```
"by the entity value" - what do you mean? If user gave a `name`, what do we query here - the ID?

e0ipso replied:
The comment is a bad one. I'll fix it, what I meant is:

_If referencing an entity by an alternate ID, retrieve the actual Drupal's entity ID using getReferencedId_

amitaibu commented on src/Plugin/resource/DataProvider/DataProviderEntity.php:
```diff
@@ -592,28 +598,73 @@ protected function queryForListSort(\EntityFieldQuery $query) {
+      foreach ($target_parts as $delta => $target_part) {
+        if ($target_part != $field_name_parts[$delta]) {
+          // There is a discrepancy between target and field name.
+          throw new BadRequestException(sprintf('The target "%s" should be a part of the field name "%s".', $filter['target'], $filter['public_field']));
```
"target" isn't very clear. Maybe `sub-field`?

e0ipso replied:
Actually I like `target`, more than `sub-field` since you are specifying what resource you are filtering.

See http://gist.github.com/mateu-aguilo-bosch/1acbced683908ff631e3.

amitaibu commented on src/Plugin/resource/DataProvider/DataProviderEntity.php:
```diff
@@ -856,4 +907,199 @@ protected function validateBody($body) {
+    else {
+      $query->propertyCondition($id_property, $value);
+    }
+    // Only one result is returned. Make sure that you are using "unique" fields
```
Make sure that you are using "unique" fields .. => maybe: `This assumes the reference fields are unique for every entity.`

e0ipso replied:
:+1:

amitaibu commented on src/Plugin/resource/DataProvider/DataProviderEntity.php:
```diff
@@ -856,4 +907,199 @@ protected function validateBody($body) {
+  }
+
+  /**
+   * Get the field name for a single item.
```
seems we return more than just the name.

e0ipso replied:
:+1:

amitaibu commented on src/Plugin/resource/DataProvider/DataProviderEntity.php:
```diff
@@ -856,4 +907,199 @@ protected function validateBody($body) {
+   * @throws \Drupal\restful\Exception\BadRequestException
+   *
+   * @return array
+   *   The result.
```
worth specifying the returned keys, as this method isn't easy to follow.

e0ipso replied:
:+1:

amitaibu commented on src/Plugin/resource/DataProvider/DataProviderEntity.php:
```diff
@@ -856,4 +907,199 @@ protected function validateBody($body) {
+   * @return array[]
+   *   An array of fields with name and type.
+   */
+  protected function getFieldsFromPublicName($name) {
```
`getFieldsFromPublicName` => `getFieldsInfoFromPublicName`?

amitaibu commented on src/Plugin/resource/DataProvider/DataProviderEntity.php:
```diff
@@ -856,4 +907,199 @@ protected function validateBody($body) {
+   */
+  protected function addNestedFilter(array $filter, \EntityFieldQuery $query) {
+    $relational_filters = array();
+    foreach ($this->getFieldsFromPublicName($filter['public_field']) as $field) {
```
$field => $field_info

amitaibu commented on src/Plugin/resource/DataProvider/DataProviderEntity.php:
```diff
@@ -856,4 +907,199 @@ protected function validateBody($body) {
+  protected function addNestedFilter(array $filter, \EntityFieldQuery $query) {
+    $relational_filters = array();
+    foreach ($this->getFieldsFromPublicName($filter['public_field']) as $field) {
+      $relational_filters[] = new RelationalFilter($field['name'], $field['type'], $field['column'], $field['destination'], $field['entity_type'], $field['bundles']);
```
where is `$field['destination']` coming from?

e0ipso replied:
It's a vestige of a failed attempt. Removing.

amitaibu commented on src/Plugin/resource/DataProvider/DataProviderPlug.php:
```diff
@@ -177,39 +176,16 @@ protected function applyFilters(array $plugins) {
-        if (!$match) {
-          // Property doesn't match the filter.
+      $interpreter = new DataInterpreterPlug($this->getAccount(), new PluginWrapper($plugin));
+      $this->fieldDefinitions->setInterpreter($interpreter);
```
I have no clue what we're doing here ;)

e0ipso replied:
Worth a comment then.

:+1:

amitaibu commented on tests/RestfulJsonApiTestCase.test:
```diff
@@ -174,35 +293,70 @@ class RestfulJsonApiTestCase extends RestfulCurlBaseTestCase {
+
+
+    // 6. Assert request processing.
+    $request = Request::create('', array(
```
nice work, that you don't have to mock all the world to test it!

e0ipso replied:
thanks! :-)

amitaibu commented:
My brain just shut down after review of 37 files :smile:

e0ipso commented:
> My brain just shut down after review of 37 files

Whoa! I can't believe that you reviewed all of the submitted PRs at once. Thanks a lot!

Commit:
Address peer review feedback

e0ipso commented:
@amitaibu I have addressed your feedback. I'll need a bit more time to check on the _base table_ thing.

_Note to self_: EFQ will use it's internal method `propertyQuery` if `$this->fields` is empty (meaning that there are no field conditions or field sorts). If `$this->fields` is not empty, then it uses the field storage query callback (`field_sql_storage_field_storage_query`). Make sure that we can join on the base entity table.

Commit:
Fix plug resource

Commit:
Add base table if necessary

e0ipso commented:
> I'll need a bit more time to check on the base table thing.

That is fixed too. Thanks for the heads up. I was aware of that, but I thought that with the way we construct queries in RESTful we would not need this. Apparently we do :-)

e0ipso commented:
And merged!

@amitaibu since you reviewed the code in this PR and this PR contained all of the others, I merged all of them.

It took me 1h 30m to do the merges due to conflict resolutions :-(

amitaibu commented:
:cake: :)
